src/Mod/Sketcher/Gui/Utils.cpp

```diff
@@ -209,6 +209,9 @@ bool SketcherGui::isSimpleVertex(const Sketcher::SketchObject* Obj, int GeoId, P
     const Part::Geometry *geo = Obj->getGeometry(GeoId);
     if (geo->getTypeId() == Part::GeomPoint::getClassTypeId())
         return true;
+    else if (geo->getTypeId() == Part::GeomBSplineCurve::getClassTypeId() &&
```

Comment:
I do not quite understand this.

IsSimpleVertex (the name may be misleading), refers to an unconnected point.

For example an unconnected point cannot be used to apply a tangency point-to-point or a perpendicular point-to-point. B-Splines admit this type of constraints (without this commit). Select one endpoint of an arc-of-circle one endpoint of a bspline and select the perpendicular or the tangency constraint. I have not compiled this PR yet, but unless I am missing something, this functionality will be gone.

Do I make sense?

Reply:
> IsSimpleVertex (the name may be misleading), refers to an unconnected point.
> 
> For example an unconnected point cannot be used to apply a tangency point-to-point or a perpendicular point-to-point.

I see. If I understand this correctly, this is what is happening:
- `isSimpleVertex` is created to reject "simple vertices" from being used in tangent/perpendicular constraints;
- knots are visualized as "simple vertices" that lie on the B-Spline;
- `isSimpleVertex` is used to accept knots in `CmdSketcherIncrease/DecreaseKnotMultiplicity::activated()`, but end points of B-splines are not accepted. 

This doesn't cause any issues in non-periodic splines since we can't practically change multiplicity of the end (i.e. first and last) knots (it is already at the highest and OCC does not allow us to decrease it). However, knot multiplicity of first and last knots in periodic spline (which are by it's nature the same) can be changed, and that is not reflected in Sketcher.

I might have to use a different function for what I intend to do in "Increase/Decrease Knot Multiplicity" command. Perhaps more appropriate would be to use `isBSplineKnot` (and make changes there to ensure end points of splines are also treated as knots). 

> B-Splines admit this type of constraints (without this commit). Select one endpoint of an arc-of-circle one endpoint of a bspline and select the perpendicular or the tangency constraint. I have not compiled this PR yet, but unless I am missing something, this functionality will be gone.
> 
> Do I make sense?

I am a little confused about how _something we cannot do with unconnected points_ relates to _something we can do with b-splines_.

Reply:
Ok. I understand your point.

I agree that knots are non-connected points (at least within the meaning of the sketcher, one could argue they are connected points in the same that they connect the piecewise parts of the B-Spline). However, it is my understanding that start and endpoint are not knots. I would say that they are coincident with knots. Endpoints do not have a InternalAlignment constraint of type knot or a extension attribute (SketchGeometryExtension, GeometryFacade) of type knot.

If you want to extend the change of multiplicity to endpoints, I see no problem in using them, as this is what the user will select, but under the understanding that we are not selecting knots, but a point coincident with them. Then we are taking this information to apply the operation to the right knot.

In this understanding, I think it is acceptable to use (to reject): isSimpleVertex || (bspline && ( start || end) ), which you may well encapsulate it in a free function (isKnotOrBSplineEnd). I would not change isBSplineKnot because of what is explained above.

Reply:
> I would suggest we try to add said constraint, but the endpoint can be common to multiple curves and that is a question we would want to address in the future (especially when OCC allows decreasing multiplicity of non-periodic end knots).

Ah, I see. We do have knots that are knots exactly at that position:
![Peek 18-02-2022 20-52](https://user-images.githubusercontent.com/5249162/154753061-ec1ef5d9-c62f-4595-b608-157a8a04dd06.gif)

I have deleted the poles at the endpoints so that it becomes clearer. There are three "simpleVertices". Two are coincident with the curve endpoints, but are rendered under the endpoints, so they are never picked. These knots do have the constraint and the extension.

You could theoretically select the knot point in the elements widget and trigger the tool (that won't work because it is unsupported by OCC, and currently also by GCS, the endpoint tangency and perpendicular will break, together with all models making use of them, as the current implementation at GCS is based on the property of non-periodic BSplines with degree multiplicity at the endpoints, that the pole is coincident with the endpoint).

The problem with coincident points, I think it is a general problem of selection which requires a solution for enhancing selection. I am unsure if RT had an improvement in his branch for this...

Reply:
> We do have knots that are knots exactly at that position: 

That is odd. I used `gdb` to run `getCoincidentPoints` (or whatever the appropriate function is called), but it returned empty. That was my first idea.

> You could theoretically select the knot point in the elements widget and trigger the tool (that won't work because it is unsupported by OCC, and currently also by GCS, the endpoint tangency and perpendicular will break, together with all models making use of them, as the current implementation at GCS is based on the property of non-periodic BSplines with degree multiplicity at the endpoints, that the pole is coincident with the endpoint).

If you want I can add another check for non-periodic b-spline that shows a Qt message saying this is not supported.

> The problem with coincident points, I think it is a general problem of selection which requires a solution for enhancing selection. I am unsure if RT had an improvement in his branch for this...

If that exists my work is kind of moot :grin:

Reply:
@realthunder in your branch is there a fix/workaround for sketcher cases where certain points (possibly different type) are inherently coincident on one another (like a bspline knot, pole and/ or end point), and an operation can only happen on one type? For example we can only increase multiplicity of knots but end points are selected.

Reply:
> The problem with coincident points, I think it is a general problem of selection which requires a solution for enhancing selection.

Do you think we could filter out certain kinds of points by making a `DrawSketchHandler` for these tools, especially after #5398?

Reply:
Yes, there is a general solution in my branch to pick coincident geometry, even constraints. I actually attemtped a PR for that a few days ago because some user requested it. But it requires a larger change then I expected, many of them are in the core actually. In addition, because of the recent huge refactor of ViewProviderSketch, those changes are not well tested. So I decided to wait. If you guys really need it now, I can try again.

Reply:
@realthunder it's not particularly urgent but good to have. This PR set out to do the exact thing but on a case by case basis. If your implementation is general it'll save quite a bit of duplication of effort.

Reply:
Maybe you can first give a try using my [release](https://github.com/realthunder/FreeCAD_assembly3/releases). If you are on Linux, you can also try the more recent [snap build](https://snapcraft.io/freecad-realthunder). And see if it covers your use case. Just use mouse move to highlight the geometry or constraint, and presse 'G, G' shortchut to bring up the geometry picker menu.

Reply:
> Maybe you can first give a try using my [release](https://github.com/realthunder/FreeCAD_assembly3/releases).

Thanks for the suggestion. I tried it and it does seem like what I need. There are some other details though:
1. There is no information of the kind of vertex (or any other entity), i.e. is it an end point, control point or knot.
2. If something is selected and a constraint is applied, We need it to make a better decision about that. Like knot multiplicity change will only happen on knot points, and we can only drag a high degree knot because it is coincident on a control point.

Reply:
@AjinkyaDahale 



> Do you think we could filter out certain kinds of points by making a `DrawSketchHandler` for these tools, especially after #5398?

Not without a substantial amount of effort. I am not aware of a way to apply a pickstyle to parts of a pointset in coin. Rather, the solution may be to get everything crossing the pick ray and allow the user to select what he means.

At DrawSketchHandler level, you could make a gate filter, like many DrawSketchHandlers use. However that would prevent from selecting certain points, without enabling you to pick points that are under.

If RT has a solution, then it makes sense to see if it can be implemented. However, we are close to a feature freeze, and I trust RT's impact assessment. If he is of the opinion that it involves substantial core changes, then it may be reasonable to wait until the next release cycle.

Reply:
> If RT has a solution, then it makes sense to see if it can be implemented. However, we are close to a feature freeze, and I trust RT's impact assessment. If he is of the opinion that it involves substantial core changes, then it may be reasonable to wait until the next release cycle.

In that case there may still be point in making the workarounds that I intend to.

Reply:
> 1. There is no information of the kind of vertex (or any other entity), i.e. is it an end point, control point or knot.

@AjinkyaDahale SketchObject can override new API [getHigherElements()](https://github.com/realthunder/FreeCAD/blob/8f08caed351bcf2e7c621bd834b931693e92c72c/src/Mod/Sketcher/App/SketchObject.cpp#L9508) to add new entries to the `Picked geometry` menu. The existing implemention checks if the picked vertex corresponds to a coincidence constraint to allow user to select those constraints instead. Same thing can be done for knots I think.

> 2\. If something is selected and a constraint is applied, We need it to make a better decision about that. Like knot multiplicity change will only happen on knot points, and we can only drag a high degree knot because it is coincident on a control point.

The above function makes it possible to select new types of elements, e.g. Knot1, you can then implement your command to check for that in selection notification.

> If he is of the opinion that it involves substantial core changes, then it may be reasonable to wait until the next release cycle.

Yeah, it does require quite some core changes. I'd suggest to wait.